doc: update ERC-7786 (#2)
defi-wonderland/ERCs: ERCS/erc-7786.md

### Message Field Encoding

A cross-chain message consists of a sender, recipient, payload, extra data, and optional hooks.

### Sender and Recipient

> The sender account (in the source chain) and receiver account (in the destination chain) MUST be input in an Interoperable Address binary format, specified in ERC-7930, which MUST be serialized according to CAIP-350. Note that the recipient field MAY opt not to contain the chain or address, but the sender MUST include it.

Reviewer: > Note that the recipient field MAY opt not to contain the chain or address, but the sender MUST include it.

is this correct? why can the recipient field opt out of both chain+address?

Author: As we talked today, I opted to 

> The recipient field MAY be omitted or zeroed (contain an Interoperable Address with all fields set to zero) to indicate an unspecified destination, such as when a broadcast mode is employed and the final recipients are determined by the protocol itself.

### Payload

The payload is an opaque `bytes` value.

### Extra Data

> The extraData is an opaque `bytes` value. This field stays OPTIONAL and MUST be interpreted only by the origin. MAY stay unused, whose case MUST be left empty.

Reviewer: ```suggestion
`extraData` is an opaque `bytes` value. This field is OPTIONAL and MUST be interpreted only on origin. If left empty this field is considered not in use.
```
not sure about `interpreted` being the right word here.

### Hook Data

The hook is a structure that invokes a low-level call in the source chain atomically after the message is sent. It is never forwarded to the destination chain as part of the core message sending flow. 

If unused, MUST be empty. It is RECOMMENDED to standardize hooks as part of new ERCs. For instance, Gateways MAY delegate hook execution to a dedicated *hook executor*. Nevertheless, implementation details are outside the scope of the standard. 

### Sending Procedure

> An **Source Gateway** is a contract that offers a protocol to send a message to a receiver on another chain. MUST implement `IERC7786GatewaySource`.

Reviewer: ```suggestion
A **Source Gateway** is a contract that uses a messaging protocol to send a message to a receiver on another chain. MUST implement `IERC7786GatewaySource`.
```

```solidity
interface IERC7786GatewaySource {
  // Optional Hook data struct
  struct HookData {
    bytes hookPayload;   // Low-level call data: selector and parameters for the hook

>     ILocalAddress hook;  // Local origin chain address

Reviewer: where is `ILocalAddress` specified? can this be just `address`?

Author: changed to `address`

    uint256 value;       // Optional, amount of native token forwarded to the hook 
  }
 
  // MessageSent event
  event MessageSent(
    bytes32 outboxId,  // Unique outbox ID (may be 0 if the gateway doesn’t track)
    bytes sender,      // Binary Interoperable Address
    bytes recipient,   // Binary Interoperable Address
    bytes payload,     // Message content
    bytes extraData,   // Optional encoded gateway metadata (leave empty if unused)
    HookData hookData  // Optional Hook data 
  );
  
  // sendMessage function
  function sendMessage(
      bytes calldata recipient,                   // Binary Interoperable Address 
      bytes calldata payload,                     // Message content
      bytes calldata extraData,                   // Optional encoded gateway metadata (leave empty if unused)
      HookData calldata hookData                  // Optional Hook data

>   ) external payable returns (bytes32 outboxId);  // Unique ID (may be 0 if the gateway doesn’t track)

Reviewer: is there any reason behind `outboxId`? could we use `messageId`?
```suggestion
  ) external payable returns (bytes32 messageId);  // Unique ID
```
also, we need to make this a requirement. it should NOT be possible to reply with id 0, it should be a valid id.

Author: This reason for leaving them as `outboxId` was based on early feedback provided by Frangio, since some protocols use TX hash + log index as the `messageId`, which is not available in the execution environment. It is assumed the protocol can generate a deterministic `messageId`, the gateway can simply return that same value as the `outboxId`

Reviewer: i mean, where is the rationale behind `outboxId` being the correct name? why not use `messageId` as the name for it?

- The protocol MUST guarantee Liveness: A sent message is eventually delivered to the destination, assuming Liveness and censorship-resistance of the source and destination chains.
- The protocol SHOULD guarantee Timeliness: A sent message is delivered at the destination within a bounded delivery time, which should be documented.
- The above properties SHOULD NOT rely on trust in some centralized actor. For example, safety should be guaranteed by some trustless mechanism such as a light client proof or attestations by an open, decentralized validator set. Relaying should be decentralized or permissionless to ensure liveness; a centralized relayer can fail and thus halt the protocol.

## Rationale

**Separation of concerns**

The interface keeps a single, predictable entry point and lets every bridge expose its peculiarities through two completely generic fields:

- `extraData` : an opaque blob the gateway consumes internally.
- `hookData`: an optional low-level call executed on the origin chain after the `sendMessage` logic.

> This is from the separation of concerns on what the Gateway specific features expose and any other external logic provided by the underlying cross-chain protocol. Both should be considered complementary extensions of the primitive core messaging.

Reviewer: the first phrase is a bit confusing & overextended. can we rephrase it?

Author: rephrased

> An application can swap one gateway address for another in storage and emit the same messages. That portability guards the app against a gateway being paused, deprecated, or simply breaking. It also enables *bridge layering*: a “meta‑gateway” can relay the same opaque message through two or more underlying bridges, choosing the combination that gives the desired blend of liveness and safety.

Reviewer: beauty!

**On Extra Data**

> Some bridges need specific parameters provided by the sender. Those parameters travel in `extraData`; if the sender leaves the blob empty, the gateway should execute the message at expected or revert otherwise, without unintended logic behavior.

Reviewer: ```suggestion
Some bridges need specific parameters provided by the sender. Those parameters are specified on the `extraData` parameter. 
```
> If the sender leaves this parameter empty, the gateway should execute the message as expected or revert otherwise, without unintended logic behavior.

this is kind of redundant, would you rather say:
"If the sender leaves this parameter empty, no additional data will be provided to the underlying messaging protocol from the perspective of the gateway. Note that additional information can always be provided to the underlying messaging protocol by other means that are considered out of the Gateway flow thus also out of the scope for this standard."

Author: solved

> Gateways MAY expose `decodeExtraData(bytes)` for off‑chain tooling, which protocols MAY opt to standardize.

Reviewer: do we want to specify this right now?

Author: I made it more lax

  hookPayload: abi.encodeCall(
    RelayerPaymentsContract.payForGas,
    (messageId, 300_000, userAddress)
  ),
  value: 0.02 ether
});
```

Since the low-level call could derive itself into other subsequent calls, hooks MAY call other hooks or have their conditional logic for sequential calls.

## Backwards Compatibility

Existing cross-chain messaging protocols implement proprietary interfaces. We recommend that protocols natively implement the standard interface defined here, and propose the development of standard adapters for those that don't.

## Security Considerations

**Handling addresses**

Interoperable Addresses (ERC‑7930) rely on CAIP‑350 serialization. Using non‑canonical encodings can lead to silent delivery failures. Gateways SHOULD reject non‑canonical encodings and MAY normalize them before emission.

**`ExtraData` parsing**

Because `extraData` is opaque to the interface, each gateway SHOULD validate its format and gracefully revert with an explicit error (e.g., `UnsupportedExtraData()`) if the blob cannot be decoded.  Never assume well‑formed input. It is RECOMMENDED to include a view function for this purpose.

**Hook Execution**

> Gateway never should call an untrusted hook directly. As it is recommended that hook calls have a dedicated executor, it is up to the protocol to ensure the hook entity is the expected one. After the call, the Gateway storage and allowances should be left untouched.

Reviewer: ```suggestion
The Gateway should never call a non-trusted hook directly. As it is recommended that hook calls have a dedicated executor middleware contract, it is up to the protocol to ensure the hook entity is the expected one.
```
